### ISSUE TITLE
Some SerialPort tests very sensitive to flow control details

### ISSUE THREAD

**Issue by willdean** [dotnet/corefx#16033]
This is so that I can ActiveIssue a bunch of failing tests as precisely as possible, and also so that I can get  my thoughts straight on this.

**tl;dr**
A significant number of the legacy SerialPort tests rely on specific traditional hardware behaviour which is not available on the CI or in the future.  We need to work around this.

**Full Details**
A lot of the serial port tests, including some of the ones which run with a single, open, serial port (e.g. on the CI) rely on a particular behaviour of the serial.sys driver / 16550 UART.   These tests arrange that hardware flow control is enabled on the port and set to block transmission, and they then write data to the port.  Because the flow control is adverse, (and because serial.sys does not rebuffer tx data within itself, nor offer it to the UART when the flow-control prohibits), this causes the data to queue-up on the user-mode side of the serial.sys (i.e. in SerialPort).  

This allows asynchronous and queue-reporting SerialPort behaviour to be tested - there is a pending write which will not complete as long as the flow control conditions remain the same.  In tests where a null modem connection is available, the test can adjust the flow control state using the other port - so it can, for example, delay the start of a transmission but then allow it to complete normally.

My feeling is that these tests are, in principle, useful - the conditions that they test are always possible in real life, it's just that they're often fleeting and occur in the middle of large data transfers - being able to create them in 'slow motion' is valuable.

**The problem**
The difficulty in 2017 is that that genuine serial.sys/16550 setups are dwindling in number - we mostly have several other things:

*USB Serial Ports* - These tend to have buffering in the driver, in the USB stack and in the silicon at the far end of the USB cable.   Hardware flow control *has* to be carried-out within the silicon, and somehow reflected back up to the user-mode level after everything's backed-up, or by using some hybrid of both blocking up the data *and* sending out of band flow-control status back to the driver to tell it to put the brakes on.  My tests with FTDI USB-Serial adaptors didn't give me terribly clear results, but it is clear that they don't instantly block writes when the flow control is adverse.  USB-Serial solutions don't have the best reputation for reliability either, although 100% fidelity emulation is impossible anyway.

*Virtual Machines* - I'm not clear what the CI hardware/software is, but it appears that we have some kind of virtualised serial port, which probably has some buffering within it.  Either way, tests that rely on this blocking behaviour tend to fail on the CI.
*Serial cards with fancier UARTS*- If you go out and buy a PCI serial port card in 2017, there's a good chance that it will have some kind of enhanced UART on it, which may well have hardware assistance for flow control - this means that the driver will transfer a certain amount of data to the UART hardware even when the flow control is adverse.

**Possible solutions**
1. Probe for this behaviour and then use additional ```ConditionalFact``` qualifiers to disable tests which rely on it.
2. Bigger writes, more relaxed queue detection.   A device which is working properly should *eventually* back-up writes if the flow control is adverse.  The test will no longer be able to verify that *all* the written data is queued-up (some of it will have gone out to the hardware), but other aspects of the test could complete.  It's possible that we might be able to probe at startup for the kind of behaviour we could expect.

All thoughts and comments welcome.

#15752, @JeremyKuhne (sorry if you've read this far)

**Comment by JeremyKuhne** [comment]
@willdean, perhaps a bit of both of your solutions? Assuming probing is possible I would think doing that and adding new tests for your second option.

I haven't looked closely at all of these, but we should make sure they're clearly documented. Assuming it isn't too difficult, tossing them all in together (class/es) might be good bucketing.

**Comment by willdean** [comment]
@JeremyKuhne I think you're right about needing a bit of both - I can certainly improve the pass rate here on 'enhanced uarts' by merely increasing the size of buffers to make sure that they exceed the capabilities of the hardware to swallow data.  Such an increase could be made across the board without any particular adverse effect on hardware which has the more traditional behaviour.

However, there are other tests which write a single byte and expect to see it blocked - tests like that are going to need to be either changed/deleted or selectively skipped on the basis of some probing.

**Comment by willdean** [comment]
This probing behaviour is now in the test suite, with a suitable call  `HasHardwareFlowControl` which can be used to attribute tests to show that they require proper flow control.
